### PR TITLE
fix wrong error for metric.structural_similarity when image is too small

### LINKED ISSUE #5366: metric.structural_similarity throws the wrong error for small sizes
## Description

when using `structural_similarity` when the images are small (<7 in any dimension) a `ValueError` gets thrown, but with the wrong text:

```
ValueError: win_size exceeds image extent.  If the input is a multichannel (color) image, set multichannel=True.
```

Instead of misleading the user into thinking there's something wrong with the way they read in, or converted, their image data, this error should read:

```
ValueError: image data too small to perform SSIM analysis on, please ensure that your images are at least 7x7.
```

## Way to reproduce
```python
import cv2
from skimage.metrics import structural_similarity

def gray(img):
	return cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

image = cv2.imread('test.jpg', cv2.IMREAD_COLOR)

# a 7x7 crop will work just fine
crop1 = image[0:7, 0:7]
crop2 = image[8:15, 8:15]
structural_similarity(gray(crop1), gray(crop2))

# a 6x6 crop will cause a ValueError, with the wrong message
crop1 = image[0:6, 0:6]
crop2 = image[8:14, 8:4]
structural_similarity(gray(crop1), gray(crop2))
```

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print("scikit-image version: {}".format(skimage.__version__))
import numpy; print("numpy version: {}".format(numpy.__version__))
```

```python
3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 10:41:24) [MSC v.1900 64 bit (AMD64)]
Windows-10-10.0.19041-SP0
scikit-image version: 0.18.1
numpy version: 1.20.2
```

### PR DESCRIPTION
**Pull request by lilisako**
## Description
- closes #5366 

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/metrics/tests/test_structural_similarity.py [hunk 1]
```diff
@@ -239,3 +240,6 @@ def test_invalid_input():
         structural_similarity(X, X, K2=-0.1)
     with testing.raises(ValueError):
         structural_similarity(X, X, sigma=-1.0)
+    # image is too small
```
Comment (mkcor) [inline skimage/metrics/tests/test_structural_similarity.py:243]: Can you also add a test for when win_size exceeds image extent? Thanks!

Comment (lilisako) [reply skimage/metrics/tests/test_structural_similarity.py:243]: @mkcor  Thanks for the review!
I think it was already added here!  https://github.com/scikit-image/scikit-image/blob/e9b086d32f6cc5271fd3c565a2a502fa74f93bd4/skimage/metrics/tests/test_structural_similarity.py#L233-L235 
If I am misunderstanding something, please let me know!

Comment (mkcor) [reply skimage/metrics/tests/test_structural_similarity.py:243]: Oh, right! My intention would be to distinguish between the two cases... maybe:
```py
    # win_size exceeds image extent                                            
    with testing.raises(ValueError) as cm:                                      
        structural_similarity(X, X, win_size=X.shape[0] + 1)
        assert 'win_size exceeds image extent' in cm.exception.msg
```
and
```py
    # image is too small                                                       
    with testing.raises(ValueError) as cm:                                     
        structural_similarity(Z, Z)                                            
        assert 'Image is too small' in cm.exception.msg
```
?

Comment (lilisako) [reply skimage/metrics/tests/test_structural_similarity.py:244]: I see! Thanks for kindly letting me know. I just pushed new commits regarding these changes.

**Comment by mkcor** [comment]
Hi @lilisako,

I've thought about the update and I wouldn't actually make this if/else distinction; it's quite artificial: You could have 6x6 or 5x5 images and compute their structural similarity, it's just that you would need to specify `win_size=5` or `win_size=3`. I think I would leave it as one error message, describing several possibilities, mentioning the multichannel case last.

Indeed,
```py
from skimage.metrics import structural_similarity
import numpy as np

im1 = np.ones((5, 5))
im2 = np.zeros((5, 5))

structural_similarity(im1, im2, win_size=3)
structural_similarity(im1, im2, win_size=5)
structural_similarity(im1, im2)  # ValueError (it is True that "win_size exceeds image extent.")
```

Therefore, I would suggest:
"win_size exceeds image extent. Either ensure that your images are at least 7x7; or pass win_size explicitly in the function call, with an odd value less or equal to the smaller side of your images. If your images are multichannel (with color channels), set channel_axis to the axis number corresponding to the channels."

Best,
Marianne

**Comment by lilisako** [comment]
Hi, @mkcor !
Sorry for being late to get back to you. I changed the error message and I explicitly added the test cases for small images. Could you re-review the code when you have time? 

Thanks in advance. 
Risako

File: skimage/metrics/_structural_similarity.py [hunk 2]
```diff
@@ -165,9 +165,14 @@ def structural_similarity(im1, im2,
...
     if np.any((np.asarray(im1.shape) - win_size) < 0):
         raise ValueError(
-            "win_size exceeds image extent.  If the input is a multichannel "
-            "(color) image, set channel_axis to the axis number corresponding "
-            "to the channels.")
+            "win_size exceeds image extent.",
+            "Either ensure that your images are ",
+            "at least 7x7; or pass win_size explicitly",
+            " in the function call, with an odd value ",
+            "less or equal to the smaller side of your ",
```
Comment (mkcor) [inline skimage/metrics/_structural_similarity.py:172]:
```suggestion
            "less than or equal to the smaller side of your ",
```

File: skimage/metrics/_structural_similarity.py [hunk 3]
```diff
@@ -165,9 +165,14 @@ def structural_similarity(im1, im2,
 
     if np.any((np.asarray(im1.shape) - win_size) < 0):
         raise ValueError(
-            "win_size exceeds image extent.  If the input is a multichannel "
-            "(color) image, set channel_axis to the axis number corresponding "
-            "to the channels.")
+            "win_size exceeds image extent.",
+            "Either ensure that your images are ",
+            "at least 7x7; or pass win_size explicitly",
+            " in the function call, with an odd value ",
```
Comment (mkcor) [inline skimage/metrics/_structural_similarity.py:171]:
```suggestion
            "at least 7x7; or pass win_size explicitly ",
            "in the function call, with an odd value ",
```
just out of consistency ;)

File: skimage/metrics/tests/test_structural_similarity.py [hunk 4]
```diff
@@ -223,6 +223,16 @@ def test_mssim_mixed_dtype():
     assert_almost_equal(mssim, mssim_mixed)
 
 
+def test_structural_similarity_small_image():
+    X = np.zeros((5, 5), dtype=np.double)
+    Y = np.zeros((5, 5), dtype=np.double)
```
Comment (mkcor) [inline skimage/metrics/tests/test_structural_similarity.py:228]:
```suggestion
    # structural_similarity can be computed for small images if win_size is
    # a) odd and b) less than or equal to the images' smaller side
```

File: skimage/metrics/tests/test_structural_similarity.py [hunk 5]
```diff
@@ -223,6 +223,16 @@ def test_mssim_mixed_dtype():
...
+    assert_equal(structural_similarity(X, X, win_size=3), 1.0)
+    assert_equal(structural_similarity(X, X, win_size=5), 1.0)
+    # when user don't specify win_size with small image
+    with testing.raises(ValueError):
+        structural_similarity(Y, Y)
```
Comment (mkcor) [inline skimage/metrics/tests/test_structural_similarity.py:233]:
```suggestion
        structural_similarity(X, X)
```

File: skimage/metrics/tests/test_structural_similarity.py [hunk 6]
```diff
@@ -223,6 +223,16 @@ def test_mssim_mixed_dtype():
     assert_almost_equal(mssim, mssim_mixed)
 
 
+def test_structural_similarity_small_image():
+    X = np.zeros((5, 5), dtype=np.double)
+    Y = np.zeros((5, 5), dtype=np.double)
+    assert_equal(structural_similarity(X, X, win_size=3), 1.0)
+    assert_equal(structural_similarity(X, X, win_size=5), 1.0)
+    # when user don't specify win_size with small image
```
Comment (mkcor) [inline skimage/metrics/tests/test_structural_similarity.py:231]:
```suggestion
    # structural_similarity errors for small images if user doesn't specify
    # win_size
```

**Comment by mkcor** [comment]
Hi @lilisako,

Thank you for your patience! I left some suggestions when reviewing your latest version. In case you're not familiar with it, let me draw your attention to the possibility of applying suggested changes in batches: https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request

**Comment by lilisako** [comment]
@mkcor 
Thanks for the review! I've just pushed the suggestions into one commit. I appreciate for taking time to correct these changes🙏

File: skimage/metrics/_structural_similarity.py [hunk 7]
```diff
@@ -165,9 +165,14 @@ def structural_similarity(im1, im2,
 
     if np.any((np.asarray(im1.shape) - win_size) < 0):
         raise ValueError(
-            "win_size exceeds image extent.  If the input is a multichannel "
-            "(color) image, set channel_axis to the axis number corresponding "
-            "to the channels.")
+            "win_size exceeds image extent.",
```
Comment (mkcor) [inline skimage/metrics/_structural_similarity.py:168]:
```suggestion
            "win_size exceeds image extent. ",
```

**Review by mkcor — APPROVED** [review]

File: skimage/metrics/tests/test_structural_similarity.py [hunk 8]
```diff
@@ -223,6 +223,18 @@ def test_mssim_mixed_dtype():
     assert_almost_equal(mssim, mssim_mixed)
 
 
+def test_structural_similarity_small_image():
+    X = np.zeros((5, 5), dtype=np.double)
```
Comment (mkcor) [inline skimage/metrics/tests/test_structural_similarity.py:227]:
```suggestion
@pytest.mark.parametrize('dtype', [np.float16, np.float32, np.float64])
def test_structural_similarity_small_image(dtype):
    X = np.zeros((5, 5), dtype=dtype)
```
in light of #5220 /cc @grlee77 
@lilisako please merge `main` (into this branch) once #5220 is merged into `main`

Comment (mkcor) [reply skimage/metrics/tests/test_structural_similarity.py:227]: #5220 is now merged into `main` :)

Comment (lilisako) [reply skimage/metrics/tests/test_structural_similarity.py]: Thanks for letting me know! I just merged main into this branch and waiting for tests to pass!

**Comment by alexdesiqueira** [comment]
Thank you @lilisako!

File: skimage/metrics/_structural_similarity.py [hunk 9]
```diff
@@ -166,9 +166,14 @@ def structural_similarity(im1, im2,
...
-            "win_size exceeds image extent.  If the input is a multichannel "
-            "(color) image, set channel_axis to the axis number corresponding "
-            "to the channels.")
+            "win_size exceeds image extent. ",
+            "Either ensure that your images are ",
+            "at least 7x7; or pass win_size explicitly ",
+            "in the function call, with an odd value ",
+            "less than or equal to the smaller side of your ",
+            "images. If your images are multichannel ",
+            "(with color channels), set channel_axis to ",
```
Comment (Pomax) [inline skimage/metrics/_structural_similarity.py:175]: small note. but: python does not need commas here.

Comment (mkcor) [reply skimage/metrics/_structural_similarity.py:175]: Very true, good catch! Waiting for your PR to fix this :smile:

Comment (Pomax) [reply skimage/metrics/_structural_similarity.py:175]: See the issue this closes as for why I won't be submitting that.

Comment (lilisako) [reply skimage/metrics/_structural_similarity.py:175]: Thanks @pomax for pointing out the issue. I fixed this in newly created PR here https://github.com/scikit-image/scikit-image/pull/5480 .

**Pull request merged**